### PR TITLE
Add support for insertion validation

### PR DESCRIPTION
**Pull request by genlu**
This would queue a signed build from merging a PR merge branch into the specified base branch, if a PR number is provided.

Here's how it works:
1. Go to AzDO and queue a new signed build
   - set 'Branch/tag' to the insertion base branch (e.g. if you PR is targeting master, but VS insertion is done from master-vs-deps, you want to set this to master-vs-deps)
   - set `PrNumber` to the PR containing the change to be validated 
2. The pipeline would automatically merge from `refs/pull/$(PrNumber)/merge` branch. If `PrNumber` not set, it will just proceed as a normal signed build; if set to invalid PR number or there's merge conflict, build would fail immediately.
3. It would also tag regular build with "OfficialBuild" and validation build with "PrValidationBuild", post-build tasks (validations and publishing) are bypassed by validation build.

todos:
1. Bypass publishing steps once validation is done (right now all post-build tasks are bypassed, including validation)
2. Create a draft VS insertion for PR validation build. Right now, the insertion pipeline for the base branch would trigger, which creates a regular PR which can't be distinguished from real insertion. I will probably just create another insertion pipeline that triggered by build with "PrValidationBuild" tag (also change existing ones to be triggered on "OfficialBuild" tag)

### CODE REVIEW

File: azure-pipelines-official.yml [hunk 1]
```diff
@@ -37,6 +38,19 @@ stages:
     - powershell: Write-Host "##vso[task.setvariable variable=SourceBranchName]$('$(Build.SourceBranch)'.Substring('refs/heads/'.Length))"
       displayName: Setting SourceBranchName variable
 
+    - powershell: |
+        Write-Host "Setting up the build for PR validation by merging refs/pull/$(PrNumber)/merge into $(SourceBranchName)."
+        if ($SourceBranchName -notlike '*-vs-deps') {
+          Write-Host  "##vso[task.LogIssue type=warning;]The base branch for insertion validation is $(SourceBranchName), which is not a vs-deps branch."
+        }
+        git pull origin refs/pull/$(PrNumber)/merge
```
Comment (JoeRobich) [inline azure-pipelines-official.yml]: In this context, does `git pull` perform a merge?

Comment (genlu) [reply azure-pipelines-official.yml]: I think it does. See https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/results?buildId=3518110

**Comment by genlu** [comment]
> Bypass publishing steps once validation is done

@jaredpar what's  the proper way to do this? Just change `eng\common\templates\post-build\post-build.yml` directly?

**Comment by genlu** [comment]
> Create a draft VS insertion for PR validation build
@JoeRobich @RikkiGibson Do you know how to set the release variable for insertion pipeline based on build tags? I'm think about tagging validations build to avoid creating regular insertion PRs.

**Comment by JoeRobich** [comment]
> Create a draft VS insertion for PR validation build

Set these to true

image

File: azure-pipelines-official.yml [hunk 2]
```diff
@@ -37,6 +38,30 @@ stages:
     - powershell: Write-Host "##vso[task.setvariable variable=SourceBranchName]$('$(Build.SourceBranch)'.Substring('refs/heads/'.Length))"
       displayName: Setting SourceBranchName variable
 
+    - powershell: Write-Host "##vso[task.setvariable variable=IsPrValidationBuild]($(PrNumber) -ne '') -and ($(PrNumber) -ne 'default')"
+      displayName: Setting IsPrValidationBuild variable
```
Comment (tmat) [inline azure-pipelines-official.yml]: >Pr [](start = 29, length = 2)

nit: `Pr` casing is odd. The usual guidelines is both capital for two letter abbrev. `IsPRValidationBuild`, `PRNumber`

File: azure-pipelines-official.yml [hunk 3]
```diff
@@ -37,6 +38,30 @@ stages:
...
+
+    - task: tagBuildOrRelease@0
+      displayName: Tag build
+      inputs:
+        type: 'Build'
+        tags: if (variables['IsPrValidationBuild']) { 'PrValidationBuild' } else { "OfficialBuild" }
+
+    - powershell: |
+        if ($SourceBranchName -notlike '*-vs-deps') {
```
Comment (tmat) [inline azure-pipelines-official.yml]: >if ($SourceBranchName -notlike '*-vs-deps') { [](start = 8, length = 45)

Could you extract this to a ps1 file in eng dir? That way one can run it manually e.g. for testing purposes

File: azure-pipelines-official.yml (hunk 2 above)
Comment (genlu) [reply azure-pipelines-official.yml]: Done.

File: azure-pipelines-official.yml (hunk 3 above)
Comment (genlu) [reply azure-pipelines-official.yml]: Sure, will do.

**Comment by genlu** [comment]
@JoeRobich @tmat Please take another look

File: azure-pipelines-official.yml [hunk 4]
```diff
@@ -45,21 +45,13 @@ stages:
...
-        }
-        Write-Host "Setting up the build for PR validation by merging refs/pull/$(PRNumber)/merge into $(SourceBranchName)..."
-        git pull origin refs/pull/$(PRNumber)/merge
-        if (!$?) {
-          Write-Host "##vso[task.LogIssue type=error;]Merging branch refs/pull/$(PRNumber)/merge failed."
-          exit 1
-        }
-        Write-Host "Getting the hash of refs/pull/$(PRNumber)/head..."
-        Write-Host ((git ls-remote origin refs/pull/$(PRNumber)/head) | Out-String)
+        tags: if (variables['IsPRValidationBuild']) { 'PRValidationBuild' } else { "OfficialBuild" }
```
Comment (tmat) [inline azure-pipelines-official.yml]: I did not notice it before - this needs to be in `${{` .. .`}}`

File: azure-pipelines-official.yml [hunk 5]
```diff
@@ -45,21 +45,13 @@ stages:
       displayName: Tag build
```
Comment (tmat) [inline azure-pipelines-official.yml]: I think 

```
 - powershell: Write-Host "##vso[task.setvariable variable=IsPRValidationBuild]($(PRNumber) -ne '') -and ($(PRNumber) -ne 'default')"
```

is missing `$(`...`)`

Should be 
```
 - powershell: Write-Host "##vso[task.setvariable variable=IsPRValidationBuild]$(($(PRNumber) -ne '') -and ($(PRNumber) -ne 'default'))"
```

**Pull request merged**